Hooks for C++ API

malvika2147 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24393 Hooks for C++ API**

Summary:
Ability to register hook on a variable, similar to python autograd API. register_hook will take a function as argument and create a CppFunctionPreHook similar to PyFunctionPreHook.
It will return the index of the hook which can be passed to remove_hook to disable the hook.

Test Plan:
Added tests.

Differential Revision: [D16861722](https://our.internmc.facebook.com/intern/diff/D16861722)

Commit:
Hooks for C++ API

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

ezyang commented:
Tests? :)

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -291,6 +292,22 @@ struct TORCH_API Variable : public at::Tensor {
+  hook_return_void_t<T> register_hook(T&& hook);
+  // Register a hook with variable return value
+  template <typename T>
+  hook_return_var_t<T> register_hook(T&& hook);
```
If my hook returns an undefined tensor, what happens?

ezyang replied:
It looks like it errors. But maybe you should treat that as "no change requested"? (In analogy to returning `None` from a normal hook.)

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -291,6 +292,22 @@ struct TORCH_API Variable : public at::Tensor {
+  template <typename T>
+  hook_return_var_t<T> register_hook(T&& hook);
+
+  void create_cpp_hook();
```
This method probably shouldn't be public?

ezyang commented on torch/csrc/autograd/cpp_hook.cpp:
```diff
@@ -0,0 +1,49 @@
+#include <torch/csrc/autograd/custom_function.h>
+
+using torch::autograd::Variable;
+static void check_single_result (Variable value, Variable result, std::string hook_name);
```
FYI it's more idiomatic in C++ to put static functions in an anonymous namespace

ezyang commented on torch/csrc/autograd/cpp_hook.cpp:
```diff
@@ -0,0 +1,49 @@
+
+  for (unsigned i = 0; i < hooks_->size(); ++i) {
+    auto &hook = (*hooks_)[i];
+    if (hook == NULL) {
```
nit: nullptr

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -100,6 +100,33 @@ variable_list _wrap_outputs(const variable_list &input_vars,
+    throw std::runtime_error(ss.str());
+  }
+
+  if (original.is_cuda() != result.is_cuda()) {
```
Hmm... where did you get this code from :o) Let's compare `device()` instead; that'll be more robust to non-CUDA/CPU things.

ezyang replied:
Check that the `dtype()` is the same, that the `device()` is the same and the `layout()` is the same.

malvika2147 replied:
This was just moved from python_hook.cpp

ezyang replied:
OK. We can be semantics preserving for now, but that code should get fixed :) (File a bug, mayhap?)

ezyang commented on torch/csrc/autograd/cpp_hook.h:
```diff
@@ -0,0 +1,17 @@
+  CppFunctionPreHook(const std::shared_ptr<hooks_list> &hooks, int value_idx);
+  variable_list operator()(const variable_list& values) override;
+
+  std::shared_ptr<hooks_list> hooks_;
```
Is there any reason this has to be a shared pointer to a vector, as opposed to just a vector?

malvika2147 replied:
The variable and all the CppFunctionHooks need to be able to access it. This is similar to what we do in python with the backward_hooks dict.

ezyang commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -180,4 +180,21 @@ void Variable::rebase_history(Edge gradient_edge) {
+
+void Variable::remove_hook(unsigned pos) {
+  auto &list = get_autograd_meta()->cpp_hooks_list;
+  TORCH_CHECK(list && pos < list->size() , "Invalid idx");
```
A more descriptive error message will be appreciated by users all around the world :)

ezyang reviewed (dismissed):
Looks legit but we shouldn't land this without tests

Commit:
Update on "Hooks for C++ API"


Summary:
Ability to register hook on a variable, similar to python autograd API. register_hook will take a function as argument and create a CppFunctionPreHook similar to PyFunctionPreHook.
It will return the index of the hook which can be passed to remove_hook to disable the hook.

Test Plan:
Added tests.

ezyang commented:
Needs a rebase past master failure

ezyang commented on test/cpp/api/autograd.cpp:
```diff
@@ -432,6 +432,79 @@ TEST(CustomAutogradTest, ReentrantPriority) {
+  });
+
+  Variable z = x * x + x * 2 + x * y + y;
+  x.register_hook(std::bind(bw_hook, 0 ,std::placeholders::_1));
```
Unless necessary (e.g., moving arguments into a lambda), it's generally better not to use `std::bind`, and just pass in a lambda that does the binding.

Commit:
Update on "Hooks for C++ API"


Summary:
Ability to register hook on a variable, similar to python autograd API. register_hook will take a function as argument and create a CppFunctionPreHook similar to PyFunctionPreHook.
It will return the index of the hook which can be passed to remove_hook to disable the hook.

Test Plan:
Added tests.
